src/Components/test/testassets/Ignitor/Program.cs

```diff
@@ -0,0 +1,142 @@
+            Console.CancelKeyPress += (sender, e) => { program.Cancel(); };
+
+            await program.ExecuteAsync(uri);
+            return 0;
+        }
+
+        public Program()
+        {
+            CancellationTokenSource = new CancellationTokenSource();
+            TaskCompletionSource = new TaskCompletionSource<object>();
+
+            CancellationTokenSource.Token.Register(() =>
+            {
+                TaskCompletionSource.TrySetCanceled();
+            });
+        }
+
+        private CancellationTokenSource CancellationTokenSource { get; }
+        private CancellationToken CancellationToken => CancellationTokenSource.Token;
+        private TaskCompletionSource<object> TaskCompletionSource { get; }
+
+        public async Task ExecuteAsync(Uri uri)
+        {
+            var httpClient = new HttpClient();
+            var response = await httpClient.GetAsync(uri);
+            var content = await response.Content.ReadAsStringAsync();
+
+            // <!-- M.A.C.Component:{"circuitId":"CfDJ8KZCIaqnXmdF...PVd6VVzfnmc1","rendererId":"0","componentId":"0"} -->
+            var match = Regex.Match(content, $"{Regex.Escape("<!-- M.A.C.Component:")}(.+?){Regex.Escape(" -->")}");
+            var json = JsonDocument.Parse(match.Groups[1].Value);
+            var circuitId = json.RootElement.GetProperty("circuitId").GetString();
+
+            var builder = new HubConnectionBuilder();
+            builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IHubProtocol, IgnitorMessagePackHubProtocol>());
+            builder.WithUrl(new Uri(uri, "_blazor/"));
+            builder.ConfigureLogging(l => l.AddConsole().SetMinimumLevel(LogLevel.Trace));
+            var hive = new ElementHive();
+
+            await using var connection = builder.Build();
+            await connection.StartAsync(CancellationToken);
+            Console.WriteLine("Connected");
+
+            connection.On<int, string, string>("JS.BeginInvokeJS", OnBeginInvokeJS);
+            connection.On<int, int, byte[]>("JS.RenderBatch", OnRenderBatch);
+            connection.On<Error>("JS.OnError", OnError);
+            connection.Closed += OnClosedAsync;
+
+            // Now everything is registered so we can start the circuit.
+            var success = await connection.InvokeAsync<bool>("ConnectCircuit", circuitId);
+
+            await TaskCompletionSource.Task;
+
+            void OnBeginInvokeJS(int asyncHandle, string identifier, string argsJson)
+            {
+                Console.WriteLine("JS Invoke: " + identifier + " (" + argsJson + ")");
+            }
+
+            void OnRenderBatch(int browserRendererId, int batchId, byte[] batchData)
+            {
+                var batch = RenderBatchReader.Read(batchData);
+                hive.Update(batch);
+
+                // This will click the Counter component repeatedly resulting in infinite requests.
+                _ = ClickAsync("thecounter", hive, connection);
```

Comment:
@javiercn / @pranavkm this causes rendering to re-render infinitely. It illustrates how to click though 😄

Reply:
So to write your perf tests you'll have to modify this `Program.cs` (which takes a URL) and then if you want to see the model that's generated on the hive you can inspect its `SerializedValue`.